### ISSUE TITLE
Improve Remote Mining

### ISSUE THREAD

**Issue by whardeman** [ScreepsOCS/screeps.behaviour-action-pattern#471]
Let's discuss the next round of improvements once #466 is merged.

Here are some starter topics, many of which just need a little attention or a decision:
* ~remote workers should only spawn if there are construction sites, and should fallback to hauler (or recycle when done)~
* ~replace the hauler base body with  [work,carry,move] if any roads in target room are visible. Tune setups to make this work.~
* hauler should flee from source when idle (evade idle?) (or possibly never idle, simply run back home)
* refactor spawn queue memory to account for users manually clearing the spawn queue
* other spawn refactor problems (priority inversion)
* ~make remote hauler setup more easily configurable to user-preference with regard to ON-ROAD or OFF-ROAD remote travel.~

**Comment by karlthepagan** [comment]
IMO first priority:
* if roads exist inside the target room then spawn road maintaining haulers
* if any CS (construction site) exists within range(3) of the remote miner, he should work them whenever his energy > 80%
* if there is a container next to the mine, make sure ALL miners sit on it
* if there is no container next to a remote mine, place a CS
* if any CS exists outside range(3) of the miner build and send a remote worker
* if any roads exist in the target room, change the hauler spec to default (uses & repairs roads)

Existing config `REMOTE_WORKER_MULTIPLIER` influences this algorithm. User can override with pioneers.

IMO next priority
* calculate maxweight for haulers based on distance and energy density
* delay building haulers until the container is finished (nearly?)

**Comment by Dasku** [comment]
Would also be nice if a hauler decided when at home if it can make the next journey, if not, stick around in home room and send a new one (so the energy isn't just dropped and lost when the creep dies on the way back)

**Comment by whardeman** [comment]
@Dasku Yes, and I might favor to immediately spawn a new hauler and recycle the old creep.

**Comment by Dasku** [comment]
Another thing:

SKK room mining may need to trigger differently:
* No claimer
* Bigger miners (4k per 300 ticks)
* Possible hauler scale changes for centre rooms

**Comment by cyberblast** [comment]
> * if roads exist inside the target room then spawn road maintaining haulers 

Why not keep it simply and always spent one WORK part for the hauler. There will always be roads somewhere. At least in the home room. There may be a short time when there are no roads in the target room. But I don't think it's worth to work around that. 

> * if there is a container next to the mine, make sure ALL miners sit on it 

Why? It's sufficient to be adjacent... (like current home-miners do) 

> * delay building haulers until the container is finished (nearly?) 

What about a two stages system. 
Stage 1: send privateers to set everything up. Build containers & roads. Finally recycle.  
Stage 2: send miner. set a timeout ( like + 50 ticks or something ) when to start building haulers...

**Comment by karlthepagan** [comment]
IMO it could just be that the SKK feature becomes an internalViral for remote mining.

**Comment by whardeman** [comment]
I understand that a number of folks are quite committed to the idea of using roads in remote mining rooms, but I don't think the code should assume or require this in order to have highly efficient remote mining out of the box. I think there's enough data to support the argument that offroad remote mining is within a few tenths of a percent as efficient as building roads to all remote rooms, and requires far less up front investment (both in terms of creeps and time).

**Comment by pandabear41** [comment]
@whardeman I think if it is stable, merge it to dev. Roads do improve energy profit by like 10%, but switching from old privateers to remoteMining will earn us 2-4x more energy depending on distance.

**Comment by Dasku** [comment]
Once its in and everyone is using it, I'm sure tweaks and improvements would be submitted quicker too.

**Comment by whardeman** [comment]
@pandabear41 @Dasku Agreed, we're really only talking about changes post PR merge. I've been using the branch extensively for days, and I feel like it's in a good place. If I could merge it right now, I would.

Regarding old privateers, yes, roads were pretty well required with the old privateers due to the creep setup WORK/CARRY/MOVE having a 1:1:1 ratio. What I'm debating is whether that should automatically apply to haulers.

**Comment by karlthepagan** [comment]
> roads were pretty well required with the old privateers due to the creep setup WORK/CARRY/MOVE having a 1:1:1 ratio

that privateer ratio is sufficient when their primary role is extracting resources from a hostile room. Maybe haulers need a behaiour + action to take energy away from creeps who are over-burdened?

**Comment by whardeman** [comment]
> that privateer ratio is sufficient when their primary role is extracting resources from a hostile room. 

@karlthepagan I'm not making the connection there. Are you talking about a room that has immediate threats from other players, thus the privateers need to harvest quickly and leave?

**Comment by karlthepagan** [comment]
well... that's all that I use privateers for now ;)

**Comment by cyberblast** [comment]
We should really try to keep it simple. 
That also means we need to agree on a common direction. 
I agree, there are pros and cons for and against roads. But it's simply easier to assume we are building roads. And the majority likes roads. 
And that doesn't mean you can't modify things individually. One likes bigger creeps, the other one likes faster ones without roads. Thats fine and should be possible and I think today we really made it quite easy to have your individual modifications. And it is intended to have them. 
But we should focus on developing a robust and clean solution, which will fit for most people. 

About this SK/NON-SK: 
I don't think we need different flags here. It's quite simple to see and differentiate by room name if it is an SK-mining-request or not. The super task will trigger different events in that case.

**Comment by whardeman** [comment]
@cyberblast Well, I'm for simplicity as well. I suppose your statement that the majority likes roads is the real reason it's simpler to assume we are building them. 

I've been trying to get people to acknowledge it as an open question that deserves careful analysis which could deeply benefit the entire alliance well into the future, but, I don't like sounding like a broken record, and I definitely don't want to foster disunity. So...

It sounds like you're making the call, so I'll concede and find a way to implement the changes in my overrides. Thanks for letting my voice my opinion.

**Comment by karlthepagan** [comment]
If it's possible for you to implement these in your overrides that will be a very strong solution from an API design perspective.

**Comment by cyberblast** [comment]
Every voice is welcome. And I am very happy about such diversity and different opinions. And I'm with you that each decision should be open and needs its analysis. But there seems to be a common opinion already. 
Which may still get discussed, what we did to some extend, but we also need to get on and I think we are at a point where a real change of that direction requires some evidence. Some effort to do such analysis to compare those different approaches.

**Comment by whardeman** [comment]
@cyberblast Yes, you're right. @karlthepagan and I had a long discussion about the data in Slack, but that's not exactly the best place to do scientific analysis, nor can people easily review it critically. If I can manage to compile my findings in a more compact format, I'll present them at such time.

**Comment by karlthepagan** [comment]
Adding 1 work part to remote haulers makes a good demo of how to viral Creep.setup.remoteHauler

I'll add that to my overrides example.

**Comment by Dasku** [comment]
Not sure if this was discussed but how come there isn't 1 CARRY part on miners so they can repair the containers?

**Comment by webremote** [comment]
Awesome work on the remoteMiner! Is it a bug that the miner doesn't automatically find the next available source, or should i set two flags?

**Comment by Methel** [comment]
With the miners now repairing containers and nearby roads, I think it might make sense to add another work part to the base body. 
That way it can still mine out the source even if it uses some time on repair.

**Comment by cyberblast** [comment]
hm yes I'm also unhappy about that source energy loss. Even more as it would fit 100% normally. 
Adding another work part (+1move) could be a solution. And I think it's worth it.

**Comment by Dasku** [comment]
This may well be a very low priority "nice to have" and not sure if its even possible however my idea would be to allow rooms to "share" a 2 source mining room. 

Below I have a mock up of my current mining setup for each room. I made my red room 3 range from the yellow boarder room on purpose to prevent it being greedy and mining the yellow room too.

However, My blue room and Pink room "share" the Yellow mining room, but that isnt the case, the tasks always end up with the pink room and have stayed there since. So now the pink room always mines the Yellow room. (in this diagram it isnt so bad as the Pink room needs the extra mining sources since it lost a 2 source room to the new White room)

My question is, could we give 1 source to Blue room and 1 source to Pink room if they are the same range from the room being mined (yellow)? Could mining tasks be allocated to rooms based on how many sources they already have access to to prevent "greedy" rooms?

![image](https://cloud.githubusercontent.com/assets/22677773/21990142/9b4c2380-dc05-11e6-8e87-5e73e12ec71e.png)

**Comment by cyberblast** [comment]
The new room selection function would select different rooms, depending on RCL, range and queue length. So that would give more load to the stronger and more idle rooms. That could also mean, the miners come from blue and pink and the haulers from white and red. 

I think thats fine so far. 

But it's a different question where they carry that energy to ... 

Currently its the haulers home room. But we should maybe change that. Prefering lower RCL rooms or rooms with lower storage level...

**Comment by karlthepagan** [comment]
I'm in favor of using commands/room to save an override to spawn/room selection. Maybe this would just result in a medium-sized hint towards the selected room just in case there's a failure.

**Comment by Dasku** [comment]
@cyberblast agreed, only where the energy is hauled to matters, not where the creep was born. I like the idea of supplying lower RCL or less energy rooms if that doesn't cause a drop in efficiency if it suddenly has to take the energy 4 range from the mining room. 

Commands or overrides is an interesting idea but adds a lot of complication? Should we try to automate a list of scenarios for the code to choose from as the "best" thing to do in each situation?

**Comment by karlthepagan** [comment]
Fully automating all of these edge cases is a monumental undertaking, but that's something to work toward. IMO letting the user specify hints to the engine will both keep the short-term goal achievable and allow for user inputs to correct for edge cases in our algorithm (edge cases that are just going to turn into more code).

**Comment by Dasku** [comment]
Good point, I wasn't even sure if something like this was either a) achievable or b) worth it

**Comment by cyberblast** [comment]
(as always) I'd like to strive to a robust and stable basic algorithm first. This may include a better basic decision making where to bring that energy. User input/configurability may come on top.

**Comment by Dasku** [comment]
Another easy one can't believe no one has mentioned yet - sending mineralMiners and another hauler when there is an extractor and minerals > 0

**Comment by pandabear41** [comment]
I agree with cyberblast on a stable remoteMining algorithm. 
Need to automatically calculate how many CARRYs needed depending on the source distance and if roads are built or not. The REMOTE_HAULER_MULTIPLIER just doesn't cut it in terms of stability. A remoteMining room one away from spawn will need less haulers than a room two away from spawn

**Comment by Dasku** [comment]
Wouldn't count * range be a ok for now? after 3 range and 3 haulers it becomes less profitable

**Comment by pandabear41** [comment]
That might be viable, but if you are RCL 4-5 it can be 6 haulers or 4 haulers depending oh how many CARRY parts your room can spawn with them.

**Comment by Dasku** [comment]
(count *range) = count; 

I don't think it would fix the problem thats why I deleted my comment, quickly changed my mind :)

**Comment by Dasku** [comment]
Came across another interesting behaviour of mining today.

![image](https://cloud.githubusercontent.com/assets/22677773/22208000/345b1c5c-e179-11e6-99d7-b1f9e22ba05d.png)

The red room is RCL7, the blue room is RCL5. When it comes to passing the mining task to a room, could we ensure it prioritises adjacent rooms always? For some reason the red room is sending haulers and miners to the mining room 1 and even caught it once sending them to mining room 2 (which is 7 range away). 

I guess this wouldn't be an issue if the haulers took the energy to the blue room, however in the meantime could we ensure adjacent rooms always get priority for miners/haulers? This should avoid their energy income being stolen by greedy rooms.

**Comment by whardeman** [comment]
I was looking at a similar issue with some of my rooms recently. I think the ROUTE_ROOM_COST feature that was developed a while back was really the right solution for this. In your case, significantly raising the cost of the room SW of the red room should achieve the desired result, but I'm not sure that feature was totally finished. @karlthepagan ?

**Comment by karlthepagan** [comment]
@whardeman ROUTE_ROOM_COST will influence which rooms spawn for a flag, but without pathfinding precalc it will not affect actual movement. It does not apply to adjacent rooms so that's one way to do it.

Also look into Task.spawn options. Right now high RCL rooms get a big benefit to spawning.

Also I've found it useful to reassign remoteHaulers's homeRoom value. Maybe this could be a standard feature?

**Comment by karlthepagan** [comment]
a few deeper issues I'm looking at lately after implementing #617
* miners don't handoff flawlessly, they turn around when they reach the destination room because the determinatedTarget has a targetOf
* at some critical level roads need a worker spawned to repair them, here's a remote room collapse caused by a combination of miner handoff and roads falling apart: https://screeps.com/a/#!/history/E47N82?t=17055875

it might be nice to have a road worker behaviour which travels along miscelaneous roads and keeps them from dying off, but this might be resolved by detouring a hauler once in a while

**Comment by chobobobo** [comment]
This thread is getting long, perhaps we should consolidate the features that have been added, and those that are remaining.  @whardeman pointed out an important change for remote haulers and miners when they arrive at the target room with nothing to do, they should pick an appropriate source to wait next to rather than wait at the border.  @karlthepagan suggested an interruptible idleMove that walks towards a target until they have something to do.  It would also need to split the arriving creeps so that they don't all sit next to the nearest source.

**Comment by karlthepagan** [comment]
I think most of our original goals are in striking distance. So I've put this in 0.7.0

**Comment by chobobobo** [comment]
If we're going to aim for 0.7.0 we should consolidate the remaining issues into a new issue with clear goals, I believe this issue has had a good life.

**Comment by chobobobo** [comment]
Did some profiling of task.mining.checkForRequiredCreeps:
counting creeps and validation: 0.5 - 2.5CPU
hauler spawning: 0.5 - 2CPU when run
worker/miner spawning: 0.5

**Comment by chobobobo** [comment]
CPU usage has been resolved by https://github.com/ScreepsOCS/screeps.behaviour-action-pattern/pull/1196
This is closed in favour of https://github.com/ScreepsOCS/screeps.behaviour-action-pattern/issues/1256
